libraries/chain/controller.cpp

```diff
@@ -3176,10 +3277,28 @@ struct controller_impl {
                report_block_header_diff(*b, ab.header());
 
                // this implicitly asserts that all header fields (less the signature) are identical
-               EOS_ASSERT(producer_block_id == ab.id(), block_validate_exception, "Block ID does not match",
+               EOS_ASSERT(producer_block_id == ab.id(), block_validate_exception, "Block ID does not match, ${producer_block_id} != ${validator_block_id}",
                           ("producer_block_id", producer_block_id)("validator_block_id", ab.id()));
             }
 
+            // verify received finality digest in action_mroot is the same as the computed
+            if constexpr (std::is_same_v<BSP, block_state_ptr>) {
+               if (!bsp->valid) { // no need to re-validate if it is already valid
+                  block_state_ptr parent_bsp = fork_db.apply_s<block_state_ptr> ([&](const auto& forkdb) {
+                      return forkdb.get_block(bsp->previous(), include_root_t::yes);
```

Comment:
Blocks are not stored in `forkdb` during replay.

Reply:
Thanks. Change to use `chain_head` as the parent of the block being built.